src/Gui/NotificationBox.cpp

```diff
@@ -0,0 +1,352 @@
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef _PreComp_
+# include <memory>
+# include <mutex>
+# include <QApplication>
+# include <QAction>
+# include <QActionEvent>
+# include <QDesktopWidget>
+# include <QEvent>
+# include <QHBoxLayout>
+# include <QHeaderView>
+# include <QLabel>
+# include <QMenu>
+# include <QPointer>
+# include <QScreen>
+# include <QStyleOption>
+# include <QStylePainter>
+# include <QStringList>
+# include <QTextDocument>
+# include <QTimer>
+#endif
+
+#include "NotificationBox.h"
+
+using namespace Gui;
+
+namespace Gui {
+
+/** Class showing the notification as a label
+ * */
+class NotificationLabel : public QLabel
+{
+    Q_OBJECT
+public:
+    // Windows implementation uses QWidget w to pass the screen (see NotificationBox::showText).
+    // This screen is used as parent for QLabel.
+    // Linux implementation does not rely on a parent (w = nullptr).
+    NotificationLabel(const QString &text, const QPoint &pos, QWidget *w, int msecDisplayTime, int minShowTime = 0);
+    ~NotificationLabel();
+    /// Reuse existing notification to show a new notification (with a new text)
+    void reuseNotification(const QString &text, int msecDisplayTime, const QPoint &pos);
+    /// Hide notification after a hiding timer.
+    void hideNotification();
+    /// Updates the size of the QLabel
+    void updateSize(const QPoint &pos);
+    /// Event filter
+    bool eventFilter(QObject *, QEvent *) override;
+    /// Returns true if the text provided is the same as the one of an existing notification
+    bool notificationLabelChanged(const QString &text);
+    /// Place the notification at the given position
+    void placeNotificationLabel(const QPoint &pos);
+
+    /// The instance
+    static NotificationLabel *instance;
+protected:
+    void paintEvent(QPaintEvent *e) override;
+    void resizeEvent(QResizeEvent *e) override;
+private:
+    /// Re-start the notification expiration timer
+    void restartExpireTimer(int msecDisplayTime);
+    /// Hide notification right away
+    void hideNotificationImmediately();
+
+private:
+    int minShowTime;
+    QTimer hideTimer;
+    QTimer expireTimer;
+};
+
+NotificationLabel *NotificationLabel::instance = nullptr;
```

Comment:
The raw pointer is the solution used in QToolTip in QT code. I was warry to change it, specially as it would mix std::unique_ptr and QT's deleteLater().

I have found this kind of solution to mix std::unique_ptr and deleteLater:
https://stackoverflow.com/questions/41402152/stdunique-ptr-and-qobjectdeletelater

I will give it a go.

Reply:
I am not pushing the changes for now, as with the changes the functional test fail (SIGENV).

It appears there needs to be both delete and deleteLater when reusing the static pointer, otherwise this:

```NotificationLabel::instance != nullptr && NotificationLabel::instance->isVisible()```

segfaults inside isVisible(), after having checked for null just before.

I will look into it tomorrow...

Reply:
Ok. I did changes in two batches and the run was without the second batch. Today I have checked the functionality and it works.